Improve handling of background IRFs with AltAz alignment

gabemery opened the pull request:
**Description**

This aims at improving the `make_map_background_irf` method when the background IRF is 3d with an AltAz alignment.
This covers the first point of issue #4860 (and the associated TODO in the code) but with a different logic than the one mentioned there.

Current limitation : the `make_map_background_irf` used to evaluate the background counts in RaDec applies a single frame transformation when using a background IRF defined in AltAz alignment. This means that if the field of view rotates during the observation, the background will be miss estimated if it is not rotation symmetric.

Proposed solution : Re-evaluate the background IRF on smaller time intervals. The original proposal was to use fixed time intervals. I have instead implemented a solution where the time intervals are created based on a maximum error on the FOV rotation.

Logic of the criteria : The rotation rate of the Radec frame in AltAz is a function of the observatory latitude, and pointing in altaz. Using this first order derivative, a given limit on rotation can be converted in a maximum observation time. It should be valid as long as the rotation rate doesn't change too fast during the output time window.

**Dear reviewer**
This pull request is ready for review. 

Should I add some documentation?

bkhelifi commented on gammapy/makers/map.py:
```diff
@@ -39,6 +39,8 @@ class MapDatasetMaker(Maker):
         This avoids extrapolation at edges and use the nearest value.
         Default is True.
+    fov_rotation_error_limit : `~astropy.units.Quantity`
+        Maximum error on the rotation angle between AltAz and RaDec frames during background evaluation.
```
Can you precise here the default value

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -98,6 +98,38 @@ def _get_fov_coords(pointing, irf, geom, use_region_center=True, obstime=None):
+    duration : `~astropy.units.Quantity`
+        Time associated with the requested rotation.
+    """
+    earth_rotation_deg_per_second = 0.004167
```
One could use instead https://docs.astropy.org/en/stable/api/astropy.time.Time.html#astropy.time.Time.earth_rotation_angle

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -98,6 +98,38 @@ def _get_fov_coords(pointing, irf, geom, use_region_center=True, obstime=None):
+        return 3600 * u.s
+    return (
+        maximum_rotation.to_value(u.deg) * np.cos(pointing_altaz.alt.rad) / denom
+    ) * u.s
```
You could handle the unit differently by giving units to `earth_rotation_deg_per_second`.
PS: @maxnoe , is there no astropy functions that provides this number?

maxnoe replied:
I think this is just `360 * u.deg / u.day`, correct?

maxnoe replied:
yes:
```

In [2]: (360 * u.deg / u.day).to(u.deg/u.s)
Out[2]: <Quantity 0.00416667 deg / s>
```

maxnoe replied:
So I'd add a constant up to:

```
EARTH_ANGULAR_VELOCITY = (360 * u.deg / u.day).to(u.deg / u.s)
```

gabemery replied:
Done: added a constant, keeping the unit as deg/day since it doesn't affect usage

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -185,11 +217,26 @@ def _map_spectrum_weight(map, spectrum=None):
     return map * weights.reshape(shape.astype(int))
 
 
+def evaluate_bkg(pointing, ontime, bkg, geom, use_region_center, obstime, d_omega):
```
Can you add a complete docstring here?

gabemery replied:
Done.
I also changed the function to be non-public.

registerrier commented on gammapy/makers/utils.py:
```diff
@@ -248,18 +294,31 @@ def make_map_background_irf(
+                ),
+                1 * u.s,
+            )
+            steptime = (
+                steptime
+                if steptime < endtime - obstime - 1 * u.s
+                else endtime - obstime
+            )
```
```suggestion
            duration = np.min(steptime, obstime)
```

gabemery replied:
`obstime` is the time corresponding to the start of the interval. It is not a duration.
The full obs duration is `ontime`.

A correct replacement would be : 
steptime = np.min(steptime, endtime - obstime)
if we ignore the minimum duration of the step that I added as 1 * u.s

registerrier reviewed (commented):
Thanks @gabemery  for tackling this issue.

I have left a number of inline comments. One question is the default value. Why 0.5 degrees? Isn't it too small? I suppose it depends on the fov and bin size no?

gabemery commented:
Thank you for the comments. I will implement them.
@registerrier The default value was not motivated. As you said it should depend on the FoV radius and bin size (reducing to the number of pixels).

A quick estimate is : 

Deviation at R for rotation of d_theta: 2 pi R * d_theta / 360 deg
size of pixel for a FOV of radius R and n_pix pixels per dimension: 2R/n_pix

If we want the deviation to be less than the size of a pixel: 
2 pi R * d_theta / 360 deg < 2R/n_pix
d_theta < 720 deg / (2 pi npix)
taking n_pix ~ 100 and using 2 pi ~ 7.2  we have
d_theta <~ 1 deg

So using this logic, 0.5 deg is not bad. But I would be OK with changing this value.

gabemery commented on gammapy/makers/utils.py:
```diff
@@ -98,6 +98,38 @@ def _get_fov_coords(pointing, irf, geom, use_region_center=True, obstime=None):
+        * np.cos(pointing_altaz.az.rad)
+    )
+    if denom == 0:
+        return 3600 * u.s
```
I wanted a way to handle the case when the denominator is 0.
Which occurs when the telescope is at one of the poles (which is constant so we could have an infinite time step),
or when the pointing az is +-90 deg. In the latter case, this is temporary so the interval needs to be limited. 

Thinking a bit more about this, one solution would be to use the derivative of this function.
The 1 hour default is okay as long as we are not very close to Zenith. 
I would be happy to have a better solution to handle this edge case.

gabemery replied:
@registerrier @bkhelifi From my tests, numpy cos function never returns exactly 0.
So these default values (changed in the latest commit) are never used and these lines can't be covered by tests. Should I remove it or keep it to be future/architecture proof?

gabemery commented on gammapy/makers/utils.py:
```diff
@@ -248,18 +294,31 @@ def make_map_background_irf(
+                else endtime - obstime
+            )
+            data += evaluate_bkg(
+                pointing, steptime, bkg, geom, use_region_center, obstime, d_omega
```
I agree, I had done it as suggested before reverting it to minimize the effect of the change when the AltAz alignment is not used.

I will re-implement this change.

gabemery replied:
I misremembered. The current implementation uses the start of the interval as reference time in both cases, as it was already implemented. (No effect in cases without rotation)

I will change `evaluate_bkg` to use the center of the time interval.

gabemery replied:
Done

gabemery commented on gammapy/makers/tests/test_utils.py:
```diff
@@ -213,6 +215,7 @@ def make_map_background_irf_with_symmetry(fpi, symmetry="constant"):
         ontime="42 s",
         bkg=bkg_3d_custom(symmetry),
         geom=WcsGeom.create(npix=(3, 3), binsz=4, axes=[axis], skydir=fpi.fixed_icrs),
+        fov_rotation_error_limit=0.5 * u.deg,
```
The function `make_map_background_irf` does not have a default value for `fov_rotation_error_limit`.
Only the class `MapDatasetMaker` has an associated default.

I would have preferred giving  `make_map_background_irf` the default and overwrite it with the value passed to `MapDatasetMaker` only if it is provided. But I didn't find an elegant way to do so.
Alternatively, I could give both a default value but I am worried it could lead to confusion later.

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -228,11 +277,8 @@ def make_map_background_irf(
-    #  the total ontime. This then handles the rotation of the FoV.
-    #  2. Use the pointing table (does not currently exist in CTA files) to
+    #  This implementation can be improved in one ways:
+    #  Use the pointing table (does not currently exist in CTA files) to
```
Indeed, for CTAO there is none. But for the HESS FITS production there is.
@LauraOlivera: is the POINTING HDU filled for HAWC/SWGO.

In Gammapy, we are currently only supporting the IACT array pointing info, ie at fixed position. This is handled with the class `FixedPointingInfo` and an `Observation` is build with such pointing class. So, the information is available (at worst, it is None).
So, either you can try to use such information in this PR, or do it into a subsequent PR.

gabemery replied:
I don't plan on handling pointing tables, as it would probably require a lot more work if the source can change.
That is why I edited the TODO to remove only the first item that is covered by this PR.

maxnoe replied:
>  POINTING HDU filled for HAWC/SWGO.

Why would it be? HAWC / SWGO are completely fixed.

bkhelifi replied:
> > POINTING HDU filled for HAWC/SWGO.
> 
> Why would it be? HAWC / SWGO are completely fixed.

In term of future formats, one should remind ourself that this HDU is not mandatory for non-pointing instrument. And then later, in our code, one will have 2 cases to handle

gabemery commented on gammapy/makers/utils.py:
```diff
@@ -98,6 +98,38 @@ def _get_fov_coords(pointing, irf, geom, use_region_center=True, obstime=None):
+    if denom == 0:
+        return 3600 * u.s
+    return (
+        maximum_rotation.to_value(u.deg) * np.cos(pointing_altaz.alt.rad) / denom
```
I will try to find a way to test it. But negative values are expected if one of the cosinus is negative. I think it can't happen for the telescope latitude (in [-90, 90] deg ), or for the pointing altitude (same range). But half of the azimuths should give negative values of `np.cos(pointing_altaz.az.rad)`.

If I am right it should not raise an error.

gabemery replied:
I tested and indeed got negative values for, as a test, observations of the galactic center from CTA northern site.
I now use absolute values for `np.cos(pointing_altaz.az.rad)`

registerrier commented:
@gabemery Is this new feature urgent or can it go into v2.0? 

Regarding the default value I hame made a few tests using your calculation of the minimum angle for 2, 1 and 0.5 degree steps with a bar like bkg IRF (0.02 deg image binning):
![2 deg step](https://github.com/user-attachments/assets/1c673a50-385e-47f4-8768-d17b374fb1f9)
![1 deg step](https://github.com/user-attachments/assets/68729623-9680-415a-9cd8-a09ec17ff8e4)
![0.5 deg step](https://github.com/user-attachments/assets/8d436c80-f328-41cb-b1e7-be4c18b547b2)

It seems to me that 1 deg still give relatively good results up to 10 deg offset.

gabemery commented:
@registerrier I don't think it is urgent. We need it in my team but we can work without a numbered release, at least for the near future.

I can change the default to 1 deg.

gabemery commented on gammapy/makers/map.py:
```diff
@@ -39,6 +39,8 @@ class MapDatasetMaker(Maker):
         This avoids extrapolation at edges and use the nearest value.
         Default is True.
+    fov_rotation_error_limit : `~astropy.units.Quantity`
+        Maximum error on the rotation angle between AltAz and RaDec frames during background evaluation.
```
Done

gabemery commented on gammapy/makers/map.py:
```diff
@@ -107,10 +109,12 @@ def __init__(
         background_oversampling=None,
         background_interp_missing_data=True,
         background_pad_offset=True,
+        fov_rotation_error_limit=0.5 * u.deg,
```
Done

gabemery commented on gammapy/makers/utils.py:
```diff
@@ -98,6 +98,38 @@ def _get_fov_coords(pointing, irf, geom, use_region_center=True, obstime=None):
     return coords
 
 
+def time_limit_rotation(maximum_rotation, pointing_altaz):
```
Done

gabemery commented on gammapy/makers/utils.py:
```diff
@@ -248,18 +294,31 @@ def make_map_background_irf(
+                time_limit_rotation(
+                    fov_rotation_error_limit, pointing.get_altaz(obstime)
+                ),
+                1 * u.s,
```
Added global value.
Splitting into two line to prevent black reformating failed.

bkhelifi commented:
Hello @gabemery, can you update your branch and fix the tests?

bkhelifi commented:
> @bkhelifi It is done. One of the tests uses a deprecated feature of not passing `obstime` which ends up as None. I added a workaround and comments with TODO to identify it.
> 
> I am unsure why "CI / ubuntu-latest, devdeps (pull_request)" fails and how to solve it. Local tests went further but still failed at scipy :
> 
> ```
> Collecting scipy>=0.0.dev0
>   Running command git clone --filter=blob:none --quiet https://github.com/scikit-hep/iminuit.git /tmp/pip-req-build-uejt8qd2
>   Running command git submodule update --init --recursive -q
>   ERROR: HTTP error 404 while getting https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/scipy/1.15.0.dev0/scipy-1.15.0.dev0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (from https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/scipy/)
> ERROR: Could not install requirement scipy>=0.0.dev0 from https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/scipy/1.15.0.dev0/scipy-1.15.0.dev0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl because of HTTP error 404 Client Error: Not Found for url: https://binstar-cio-packages-prod.s3.amazonaws.com/64774d9554a6f0c49626f52d/670259440893138e9b5b9aaa?response-content-disposition=attachment%3B%20filename%3D%22scipy-1.15.0.dev0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl%22%3B%20filename%2A%3DUTF-8%27%27scipy-1.15.0.dev0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl&response-content-type=application%2Foctet-stream&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAWUI46DZFGU7XD2YL%2F20241007%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241007T103727Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Security-Token=IQoJb3JpZ2luX2VjENP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJHMEUCIQCtOqhQgfRhUUvYJ%2BKRFUynSogLbvzUebyTCL7A6AmATAIgVzwagrszpPaCPJmoYeJUxhndrCFSwdxkVSYoEa9e4XYq%2FgQIKxAAGgw0NTU4NjQwOTgzNzgiDD6GGq72VRG9MzMQfirbBDhA3H00YC%2BVUILumVFYrKlNepONx9TVnpLhPmkiGn%2FV94ksvMusCoZ7cUKEPAW6yHGD6GDI4%2FutoHkh8CTrc%2FNXiYMcdPPWiQ%2FwQM67xXub4C%2BVMT8L3rpuWCNohmJPWkgO70csiMo0bRIbKnw4OTdLuqnQbtGageOheh%2FIowibhTnwM3HwxxqRLlMtbE%2Fq8quawB3vpmVqiLxEi9xgswePgNjSVyUL6uGg4WAKMnTK7wyPePQ3jhgRlgA5mutCNk6gFb7Ts85iZxpfPe5vdrQ5zdZSzVGF%2FS1WAqzmeJ1zPF0hxWzKW4djurFgw5iM3xACzc7AjswvUeD53DIVHCuttcZ2s17rItEMr8nwtEyL68SQcpsclrryet2ca0hPbsFAFBem5tIRojHxYN%2FPUP5XJnqVCoKVT8%2Fg6BUEWbU3HMcFRg8jkye18NXvKF%2FGHeMD6l1iuFIXEL6WQoY4yjc344HGHwVnrOhuMNw8tH%2F3nZw6ybWOrpIVegE9XWqDAczvXb5IHT7yl%2FgJZCi685Pggpg9ta7q%2FlhYEYEZEBIIozRPfD0kZxrdlDXDur43ROJAwJIZXLuK0Ty3g5BFf98yv3kM48y%2F0WdpKbRrBY9GY7coXxxMvYBhd5Y4BXxORPfmONN3zl1SyMcPjHKDkvhmoNnGayXGt%2BXh%2BDEuIHQtPiuFqC7ksaQbftKAoZpuiDR74d1dQ5oBoq7gBr0VV%2FnwjwiwCmUIAjqQV2%2ByQzBAD55oyel%2BQxnArYwp7fZsvjK7VXZbjgc9z5Lqq3O5eJF%2FSOHFA1PXNvuvdTC46Y64BjqaAWaxtu7SulQyCabeE0PKVPejTklPivU%2BMs8xZ3qlyYh1%2FZ82PU8SOaxb%2BL%2BL5IPcgYcG6WzjLEHnMdTS6onhL6PYYChFs0IwcDy%2FQbBep6Bxwvhs6Q1VGbBxdUu%2FdEXR4Uft9O3arpewD9s4F8KFKbQd0MqRQqAG%2BLlXgt9AatM6DwMyIVcz0IGL%2BhMcQR%2BmuMOHmj6xa6dzSDc%3D&X-Amz-Signature=f0cd6c5cc3dd2d6669d6b26d8cfeb4309066616a190d9694d1bb0cd63dcd6afe for URL https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/scipy/1.15.0.dev0/scipy-1.15.0.dev0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (from https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/scipy/)
> 
> devdeps: exit 1 (171.82 seconds) /home/emery/Documents/dev_gammapy> python -I -m pip install 'astropy>=0.0.dev0' 'matplotlib>=0.0.dev0' 'numpy>=0.0.dev0' 'pyerfa>=0.0.dev0' 'scipy>=0.0.dev0' git+https://github.com/scikit-hep/iminuit.git pid=209423
>   devdeps: FAIL code 1 (172.00 seconds)
>   evaluation failed :( (172.09 seconds)
> ```

An issue with the latest version of scipy. Let's wait few days.... The other tests work, so one can work on the review already;)

bkhelifi commented on gammapy/makers/map.py:
```diff
@@ -39,6 +39,10 @@ class MapDatasetMaker(Maker):
+    fov_rotation_step : `~astropy.units.Quantity`, optional
+        Maximum error on the rotation angle between AltAz and RaDec frames during background evaluation.
+        Used only when the Background IRF has an AltAz alignement.
+        Default is 0.5 deg.
```
```suggestion
        Default is 1.0 deg.
```

gabemery commented on gammapy/makers/tests/test_utils.py:
```diff
@@ -322,9 +326,29 @@ def test_make_map_background_irf_skycoord(fixed_pointing_info_aligned):
 
 
+@requires_data()
+def test_make_map_background_irf_AltAz_align(fixed_pointing_info):
```
I extended the test to check that outputs dependence on alignment and observation duration are seen.

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -185,11 +217,58 @@ def _map_spectrum_weight(map, spectrum=None):
+        geom=geom,
+        use_region_center=use_region_center,
+        #  TODO `obstime` should be provided in future versions of gammapy.
+        #  TODO Using None is deprecated. if else to be removed.
```
@registerrier : I am wondering whether one should be more strict on this, in preparation of the v2.0. The warning message in `_get_fov_coords` should mention that it will be a mandatory argument from v2.0. Any preference?

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -185,11 +217,58 @@ def _map_spectrum_weight(map, spectrum=None):
+        #  TODO Using None is deprecated. if else to be removed.
+        obstime=None
+        if obstime is None
+        else obstime + ontime / (2 * u.dimensionless_unscaled),
```
I do not understand here the logic:
"Observation time to use" + "run duration"/2. This will be outside the run duration for many steps...

gabemery replied:
The docstring may be confusing. The ontime here is the interval duration at which we evaluate the coordinates. It is steptime in the irf function. I didn't change this code section from the original version which was always applied to the full runtime. I will update the naming and docstring to make it better.

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -213,6 +292,8 @@ def make_map_background_irf(
     geom : `~gammapy.maps.WcsGeom`
         Reference geometry.
+    fov_rotation_step : `~astropy.units.Quantity`
+        Maximum rotation error to create sub-time bins if the irf is 3D and in AltAz.
```
Do you test somewhere that this is a 3D IRF? What happens in the contrary?

gabemery replied:
I test `if bkg.has_offset_axis or bkg.fov_alignment == FoVAlignment.RADEC`
which handle the 2D and 3D, RaDec aligned cases. 
Then line 333 `if bkg.fov_alignment == FoVAlignment.ALTAZ:`
handles the 3D AltAz case. Issues could arise if a 2D IRF has `(bkg.fov_alignment == FoVAlignment.ALTAZ) == True`

Looking at the current classes it should not happen randomly since the the default of fov_alignment is RaDec. But I  will add a layer of check. Using line 333 :
`if is_instance(bkg, Background3D) and bkg.fov_alignment == FoVAlignment.ALTAZ:`
or 
`if not bkg.has_offset_axis and bkg.fov_alignment == FoVAlignment.ALTAZ:`

This will allow to ignore the AltAz handling when the data irf was already evaluated as 2D.

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -185,11 +217,58 @@ def _map_spectrum_weight(map, spectrum=None):
+        If False, average over the whole region.
+        Default is True.
+    obstime : `~astropy.time.Time`
+        Observation time to use.
```
Is it not instead the start of the observation? This argument name could be very confusing

gabemery replied:
I re-used the docstring of `make_map_background_irf`. I agree it is not the most clear.
It is indeed the start of the observation.

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -185,11 +217,58 @@ def _map_spectrum_weight(map, spectrum=None):
+    pointing :  `~gammapy.data.FixedPointingInfo` or `~astropy.coordinates.SkyCoord`
+        Observation pointing.
+    ontime : `~astropy.units.Quantity`
+        Observation ontime. i.e. not corrected for deadtime
```
You mean the observation duration?

gabemery replied:
I re-used the docstring of `make_map_background_irf`. 
It is indeed the observation duration, which is also names ontime in the tables from datastores if I recall correctly.

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -248,18 +326,31 @@ def make_map_background_irf(
+            data += _evaluate_bkg(
+                pointing, steptime, bkg, geom, use_region_center, obstime, d_omega
+            )
+            obstime = obstime + steptime
```
It looks to me a bit strange to modify a function argument. I would have used a dedicated variable

gabemery replied:
Ok I will change this.

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -185,11 +217,58 @@ def _map_spectrum_weight(map, spectrum=None):
+    pointing :  `~gammapy.data.FixedPointingInfo` or `~astropy.coordinates.SkyCoord`
+        Observation pointing.
+    ontime : `~astropy.units.Quantity`
+        Observation ontime. i.e. not corrected for deadtime
```
Why the deadtime is not applied? At the end, it should be, is not it?

gabemery replied:
That is how it was previously implemented. 
Now to answer your question, I don't know. It depends on how the background IRF is defined. If it is "the number of background counts that we expect to be seen when observing a certain duration" then the answer is no.
If it is "The number of background events that could trigger the camera if it is available during a certain duration", then the livetime needs to appear at some point.

I would say the first definition is more logical. (Of course, if the ontime and the duration of GTIs are not the same, this would need to be accounted for).

bkhelifi reviewed (commented):
Thanks again for this work. I have a couple a question to understand better...

gabemery commented:
To help resuming activity on this PR, here is an overview of the remaining discussion points:

In gammapy/makers/utils.py, line 131: Handling of denominator = 0 (null rotation speed)
In my last commit I introduced the more detailled handling of the case when 1. the telescope is located such that the FoV doesn't rotate or 2. we are at the exact time when the rotation speed is null
In term of code it is :  `np.cos(pointing_altaz.location.lat.rad) == 0` or  `np.cos(pointing_altaz.az.rad) == 0`
There was a question about the default value, which I changed. Is it correct now?
I have an issue in term of test coverage. Currently, numpy.cos is never exactly equal to zero. So the lines cannot be covered in tests and would never be used in application. Should I keep it as is since it is future safe, or is it better to simplify the code?



 In gammapy/makers/utils.py, line  246 
```
obstime : `~astropy.time.Time`
        Observation time to use.
```
This is the run starting time, so it was commented that the name was confusing. While I agree, I still have it as obstime since this is what is used in `make_map_background_irf`for this parameter. To remain consistent I would change both or none. Any preferences?


I am unsure why the tests are failing for some part of the CI but not all after merging the current master.

bkhelifi commented on codemeta.json:
```diff
@@ -369,6 +369,6 @@
         "pydantic>=2.5.0",
         "iminuit>=2.8.0",
-        "matplotlib>=3.4"
+        "matplotlib>=3.4, <3.10"
```
Can you remind me why is it useful?

gabemery replied:
This was a commit  by "actions-user", I have no clue why it is done.

[8df0035](https://github.com/gammapy/gammapy/pull/5451/commits/8df00357b2cc673d013d85d30cb7b9cff6a375b2)

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -101,6 +104,40 @@ def _get_fov_coords(pointing, irf, geom, use_region_center=True, obstime=None):
+    if denom.value == 0:
+        if np.cos(pointing_altaz.location.lat.rad) == 0:
+            # Assuming that the observatory location is fixed during one observation, the time step can be infinite
+            return np.inf * u.s
```
@registerrier: should we add an info message for these 2 cases?
@gabemery: it looks that these 2 cases are not tested. Is it needed?

gabemery replied:
They are not tested because numpy.cos is never exactly zero. So in application it is currently useless.
And impossible to test. But I kept it in case numpy behaviour changes in the future or is architecture dependent. 

I am open to removing it, since I don't like having untested lines.

bkhelifi commented:
Thank a lot @gabemery. I have an open question: should we make a higher level test? e.g. getting an excess with a radial bkg and a 3D symetric bkg in AltAz... I do not know yet exactly how to do it. But it will ensure that whatever the bkg model type the results are identical when the bkg rates are

gabemery commented:
Thank you @bkhelifi ,
Couldn't we compare directly the background evaluation? The excess is not needed to know if the backgrounds are similar.

We will have to accept some differences due to the different spatial binning, but is can indeed be done.

gabemery commented:
@registerrier , after some exchanges with @bkhelifi , we arrived at some idea about the handling of the division by zero currently not covered by the tests. The issue is that numpy.cos never returns zero as far as I was able to test on my hardware. The problem with completly removing the test would be that, if future versions of numpy change the behaviour, or if it is architecture dependent, the code would fail.

One promising way to handle this would be to use a constant appearing in the computation, putting it at zero in the test. This would remove the different handling of the two cases leading to this division by zero but would prevent failure.

We also discussed the possibility of replacing the `if denominateur == 0` by  a  `try, except ZeroDivisionError` bloc.

To make it easier to compare the ideas, here is the current function : 

```
def compute_rotation_time_step(rotation, pointing_altaz):
   
    denom = (
        EARTH_ANGULAR_VELOCITY
        * np.cos(pointing_altaz.location.lat.rad)
        * np.abs(np.cos(pointing_altaz.az.rad))
    )
    if denom.value == 0:
        if np.cos(pointing_altaz.location.lat.rad) == 0:
            # Assuming that the observatory location is fixed during one observation, the time step can be infinite
            return np.inf * u.s
        if np.cos(pointing_altaz.az.rad) == 0:
            # For fixed RaDec pointing, this is temporary. A short interval is thus provided
            return 60 * u.s
    return rotation * np.cos(pointing_altaz.alt.rad) / denom
```
With this untested section: 
```
if np.cos(pointing_altaz.location.lat.rad) == 0:
            # Assuming that the observatory location is fixed during one observation, the time step can be infinite
            return np.inf * u.s
        if np.cos(pointing_altaz.az.rad) == 0:
            # For fixed RaDec pointing, this is temporary. A short interval is thus provided
            return 60 * u.s

```
Possible solutions (using EARTH_ANGULAR_VELOCITY=0 in the tests) : 

- try bloc :

```
try:
    denom = (
        EARTH_ANGULAR_VELOCITY
        * np.cos(pointing_altaz.location.lat.rad)
        * np.abs(np.cos(pointing_altaz.az.rad))
    )
    return rotation * np.cos(pointing_altaz.alt.rad) / denom
except ZeroDisvisionError:
    log.warning ?
    return 60 * u.s
```
    
    
- if statement:

```
denom = (
    EARTH_ANGULAR_VELOCITY
    * np.cos(pointing_altaz.location.lat.rad)
    * np.abs(np.cos(pointing_altaz.az.rad))
)
if denom.value == 0:
    log.warning ? 
    return 60 * u.s
return rotation * np.cos(pointing_altaz.alt.rad) / denom
```

What do you think of this?

bkhelifi commented:
It seems very nice and ready. Any other aspect to develop or test? If not, @registerrier and @QRemy could review this PR;)

bkhelifi commented:
I am wondering whether this feature should not be explained within a technical tutorial, or by updating some tutorials in a following PR

gabemery commented:
@bkhelifi I think everything that needed to be tested was covered. And all the features I wanted to include are done. So it is ready.
I will have a look at tutorials to see if it can fit somewhere.

bkhelifi reviewed (approved):
Thanks a lot. Good for me...

bkhelifi commented on gammapy/makers/utils.py:
```diff
@@ -188,11 +218,58 @@ def _map_spectrum_weight(map, spectrum=None):
+        If False, average over the whole region.
+        Default is True.
+    obstime : `~astropy.time.Time`
+        Observation time to use.
```
Add 1. what is used when it is none 2. "Default is None."

gabemery commented on gammapy/makers/utils.py:
```diff
@@ -188,11 +218,58 @@ def _map_spectrum_weight(map, spectrum=None):
+        #  TODO `obstime` should be provided in future versions of gammapy.
+        #  TODO Using None is deprecated. if else to be removed.
+        obstime=None
+        if obstime is None
```
I will remove the support of None if I am allowed. It should make things easier. (The if is not useless, I think, since this is a one line if else)

gabemery commented:
I removed the usage of obstime=None in the irf evaluation function. `_get_fov_coords` is still used in 4 other places with None so I left the deprecation warning in it.

Function
    _get_fov_coords(pointing, irf, geom, use_region_center=True, obstime=None)
Usages in All Places  (5 usages found)
    Unclassified  (5 usages found)
        gammapy  (5 usages found)
            gammapy/makers  (5 usages found)
                utils.py  (5 usages found)
                    _evaluate_bkg  (1 usage found)
                        252 coords = _get_fov_coords(
                    make_edisp_kernel_map  (1 usage found)
                        488 coords = _get_fov_coords(
                    make_edisp_map  (1 usage found)
                        439 coords = _get_fov_coords(pointing, edisp, geom, use_region_center=use_region_center)
                    make_map_exposure_true_energy  (1 usage found)
                        162 coords = _get_fov_coords(
                    make_psf_map  (1 usage found)
                        391 coords = _get_fov_coords(

registerrier reviewed (approved):
Thanks @gabemery . 

I took the liberty to change a few things:
- the times steps are computed in a specific function outside the bkg calculation. This approach can be useful for other situations e.g. for non axi-symetric effective areas.
- I re-organized a bit the tests. In particular, I changed to specific values testing instead of just checking that at least one value is different

registerrier commented:
Merging now. Thanks @gabemery 

Notes and remaining things to do:
- is the behavior completely correct in the observation consists in two different GTIs: the GTIs should be split.
- Add tests for asymmetric altaz bkg and not use_region center
- need to remove deprecation warning in `_get_fov_coord`
- This PR is breaking the API, because we need to change arguments. Functions like `make_background_map_irf` are implementation details and not part of the user API. We shouldn't have to worry about this here. This is a bit unfortunate. We should revisit what is public API and what is not.
- we should start to adapt all functions projecting IRF into maps to take `time_start` and `time_stop`
